src/Session.php

```diff
@@ -607,17 +617,50 @@ public static function loadGroups()
             ],
             'WHERE'     => [
                 Group_User::getTable() . '.users_id' => self::getLoginUserID()
-            ] + getEntitiesRestrictCriteria(
-                Group::getTable(),
-                'entities_id',
-                $_SESSION['glpiactiveentities'],
-                true
-            )
+            ] + $entity_restriction
         ]);
 
+        $is_recursive_global = Config::getConfigurationValue('core', 'use_recursive_groups');
         foreach ($iterator as $data) {
             $_SESSION["glpigroups"][] = $data["groups_id"];
+
+            // Add children groups
+            if ($is_recursive_global && $data['recursive_membership']) {
+                // Stack of children to load
+                $children_to_load = [$data["groups_id"]];
+
+                while (!empty($children_to_load)) {
+                    $next_child_to_load = array_pop($children_to_load);
+
+                    // Note: we can't use getSonsOf here because some groups in the
+                    // hierarchy might disable recursive membership for their own
+                    // children
+                    $children_data = $DB->request([
+                        'SELECT' => ['id', 'recursive_membership'],
+                        'FROM'   => Group::getTable(),
+                        'WHERE'  => ['groups_id' => $next_child_to_load] + $entity_restriction
+                    ]);
+
+                    // Iterate on the children
+                    foreach ($children_data as $data) {
+                        // Add the child to the user's groups
+                        $_SESSION["glpigroups"][] = $data['id'];
+
+                        // If the child support recursive membership, load its
+                        // children too
+                        if ($data['recursive_membership']) {
+                            $children_to_load[] = $data['id'];
+                        }
+                    }
+                }
+            }
         }
+
+        // Clear duplicates
+        $_SESSION["glpigroups"] = array_unique($_SESSION["glpigroups"]);
+
+        // Set new valid cache date
+        $_SESSION['glpigroups_cache_date'] = Config::getConfigurationValue('core', 'last_group_change');
```

Comment:
A db query could probably saved by storing current time here.
```suggestion
        $_SESSION['glpigroups_cache_date'] = $_SESSION['glpi_currenttime'];
```

Reply:
No longer relevant after switching to cache (6f7e20cf94).